driver-core/src/main/com/mongodb/internal/operation/CursorResourceManager.java

```diff
@@ -0,0 +1,277 @@
+ * others are not and rely on the total order mentioned above.
+ */
+@ThreadSafe
+abstract class CursorResourceManager<CS extends ReferenceCounted, C extends ReferenceCounted> {
+    private final Lock lock;
+    private final MongoNamespace namespace;
+    private volatile State state;
+    @Nullable
+    private volatile CS connectionSource;
+    @Nullable
+    private volatile C pinnedConnection;
+    @Nullable
+    private volatile ServerCursor serverCursor;
+    private volatile boolean skipReleasingServerResourcesOnClose;
+
+    CursorResourceManager(
+            final MongoNamespace namespace,
+            final CS connectionSource,
+            @Nullable final C connectionToPin,
+            @Nullable final ServerCursor serverCursor) {
+        this.lock = new ReentrantLock();
+        this.namespace = namespace;
+        this.state = State.IDLE;
+        if (serverCursor != null) {
+            connectionSource.retain();
+            this.connectionSource = connectionSource;
+            if (connectionToPin != null) {
+                connectionToPin.retain();
+                markAsPinned(connectionToPin, Connection.PinningMode.CURSOR);
+                this.pinnedConnection = connectionToPin;
+            }
+        }
+        this.skipReleasingServerResourcesOnClose = false;
+        this.serverCursor = serverCursor;
+    }
+
+    /**
+     * Thread-safe.
+     */
+    MongoNamespace getNamespace() {
+        return namespace;
+    }
+
+    /**
+     * Thread-safe.
+     */
+    State getState() {
+        return state;
+    }
+
+    /**
+     * Thread-safe.
+     */
+    @Nullable
+    CS getConnectionSource() {
+        return connectionSource;
+    }
+
+    /**
+     * Thread-safe.
+     */
+    @Nullable
+    C getPinnedConnection() {
+        return pinnedConnection;
+    }
+
+    /**
+     * Thread-safe.
+     */
+    boolean isSkipReleasingServerResourcesOnClose() {
+        return skipReleasingServerResourcesOnClose;
+    }
+
+    @SuppressWarnings("SameParameterValue")
+    abstract void markAsPinned(C connectionToPin, Connection.PinningMode pinningMode);
+
+    /**
+     * Thread-safe.
+     */
+    boolean operable() {
+        return state.operable();
+    }
+
+    /**
+     * Thread-safe.
+     * Returns {@code true} iff started an operation.
+     * If {@linkplain #operable() closed}, then returns false, otherwise completes abruptly.
+     *
+     * @throws IllegalStateException Iff another operation is in progress.
+     */
+    boolean tryStartOperation() throws IllegalStateException {
+        return withLock(lock, () -> {
+            State localState = state;
+            if (!localState.operable()) {
+                return false;
+            } else if (localState == State.IDLE) {
+                state = State.OPERATION_IN_PROGRESS;
+                return true;
+            } else if (localState == State.OPERATION_IN_PROGRESS) {
+                throw new IllegalStateException(MESSAGE_IF_CONCURRENT_OPERATION);
```

Comment:
This throwing does not work for the asynchronous code in `AsyncCommandBatchCursor.ResourceManager.execute`. A way to fix this is to refactor `tryStartOperation` such that it returns `Optional`/`null`-able `IllegalStateException` either when the state is not operable (the cursor was closed)<sup>1</sup> or when another operation is in progress. Then both `AsyncCommandBatchCursor` and `CommandBatchCursor` may inspect the returned value and propagate the exception using the appropriate mechanism (`throw` in the sync code, `onResult` in the async callback-based code).

P.S. One may argue that `AsyncCommandBatchCursor.next` can't actually be called concurrently with itself (there are no other operations guarded by `execute` on `AsyncCommandBatchCursor`) by a user, because we wrap it in a `Publisher`. So if we observe `localState == State.OPERATION_IN_PROGRESS` here from `AsyncCommandBatchCursor.ResourceManager.execute`, then we know there is a driver bug, i.e., we should rather have an assertion. I think, it is also a valid approach, but in this case `CursorResourceManager.tryStartOperation` has to differentiate when it's called from `CommandBatchCursor.ResourceManager.execute`, and when from `AsyncCommandBatchCursor.ResourceManager.execute`.

---

<sup>1</sup> This also requires the method to start accepting the `exceptionMessageIfClosed` parameter, similarly to `CommandBatchCursor.ResourceManager.execute`.

Reply:
Opted to throw an assertion error (added `Assertions.doesNotThrow`).